### ISSUE TITLE
nextcloud and localization features

### ISSUE THREAD

**Issue by skynick11** [Zimbra-Community/owncloud-zimlet#231]
We have two enhancements, but they are linked to each other, so I'll describe both in the same issue.

1.  Add some Nextcloud-related features (available in the Nextcloud's Administration->Sharing menu) for a link attachment:

- Set default expiration date
- Enforce password protection
- Enforce expiration date

We added three correspondent parameters into config.properties, new placeholder for the password field, and added the validation for "enforced" parameters in "send link(s)" dialogs (so, this issue is linked with #226 , question 3). 

2. Localization issues

There are several types of localization issues:
a) Some strings are hardcoded, so it's a good idea to move them into localization file(s).

b) Some strings are hardcoded concatenations of Zimbra standard messages. It may work sometimes, but it's a bad idea for non-Latin languages (at least due to changing endings of words; that's why #132 said that "GoogleTranslate gives the best result" than such concatenations; it also linked with #226 question 4).
So, we think that it's possible to replace them by new strings in the localization file(s), but temporary allow the fallback to the current behavior.

c) Some strings are Zimbra standard messages, whose are good, but in some languages the same English words can be translated by different ways due to context (or other word endings should be used due to context).
So we want to use Zimbra messages by default but allow to override the translations for some of them.

Preliminary implementation for both issues is attached: [Patches.zip](https://github.com/Zimbra-Community/owncloud-zimlet/files/4009944/Patches.zip) , patches should be applied in ascending order of numbers.

If the described/attached above is interesting, we will test the modifications more thoroughly and report the result (or apply the modified patches, if any)

**Comment by barrydegraaff** [comment]
Hello Skynick, 

Thank you for your contributions, can you send me a pull request (PR) with your 
changes when you are done testing them, instead of a zip with patches. 

To make such a PR, hit the FORK button on top of the page to copy this repo 
to your personal Github account. Then git clone your personal repo, make the changes, 
commit them back, and finally click Create Pull Request. 

That way, we can be sure all is merged in the correct order, and it makes it easier 
to rollback also. 

Thanks! 

[ https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request-from-a-fork | https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request-from-a-fork ] 
[ https://www.youtube.com/watch?v=rgbCcBNZcdQ | https://www.youtube.com/watch?v=rgbCcBNZcdQ ] 


Kind regards, 

Barry de Graaff 
Zeta Alliance 
Co-founder & Developer 
zetalliance.org | github.com/Zimbra-Community 

Signal: +31 617 220 227 
Fingerprint: 97f4694a1d9aedad012533db725ddd156d36a2d0

**Comment by skynick11** [comment]
Hello,

the changes are split into two files just for review purposes. It is better to see with them than in a single merge-commit whose changes are related to improved functionality, and whose are related to localization.

So, once the idea is approved, we begin testing.

**Comment by skynick11** [comment]
Hello,

progress at the moment
- We tested, found some bugs, fix them and test again.

- But also found two bugs in the original code:
1. The password is inserted into the mail composer's editor without html-safe encoding, which is why some characters or their combinations are treated as HTML markup. Applicable to the html editor, not applicable to the plain text editor if it's possible to use such editor in the mail composer.
2. Some passwords are truncated when sending to the server, such as "~!@#$%^&*()_+", But correctly accepted by the server in its own interface. The reason is unknown at this moment.

Which is better - to report those bugs as one or separate issues?

**Comment by barrydegraaff** [comment]
You can put them all in one PR

**Comment by skynick11** [comment]
Sorry, I don't understand future steps...

We'll create PR for new features (Set default expiration date, Enforce 
password protection, Enforce expiration date) as described in #231 after 
re-check our implementation.

But problems with htm-safe encoding of the password and a password 
truncation aren't related to this work, they are presented in the 
current state of owncloud-zimlet (we just found them).

Are you sure that it's a good idea to put bug report into the PR with 
new features, or is it better to split the PR (for #231) and the new 
issue for bugs?

03.03.2020 10:27, Barry de Graaff пишет:

**Comment by barrydegraaff** [comment]
You can put the fix for the encoding in the same pr or split them. If you do not intend to fix the issue, you can make a separate issue for it.

**Comment by skynick11** [comment]
We don't have the fix for these problems at this moment and don't know 
how to fix them too. I’m sure that we can help with testing, but so far 
I’m not sure about actually fixing bugs.

And I'm not sure that these problems are closely related to each other 
(encoding in the composer and truncating on the way to the nextcloud 
server).

Ok, I'll report bugs as a single issue, and we'll discuss the 
possibilities later.

03.03.2020 16:27, Barry de Graaff wrote:
